### PR TITLE
Make mon_data_space.py more generic (part 1)

### PR DESCRIPTION
**Pull request by EdwardSafford-NOAA**
## Description

This PR is the first step to making the file loads for the binary legacy DA monitor data files generic.  

The first cut at `data/mon_data_space.py` was able to load only one specific type of monitor files (RadMon, time-series).  This PR accomplishes the same file load into `data_collections` but does so in a generic way that will support additional DA monitors (which all use binary files but with different dimensions).

Note that in many places there is an if/elif/else on the number of dimensions in use. This will be 1, 2, or 3.  2 works with these changes; 1 and 3 are not fully formed yet.  The changes in this PR work with the existing test cases `testMonDataSpaceHirs4Metop-A.yaml` and `testMonSummary.yaml`.  I opted to halt work at this point and complete the work for dimensions 1 and 3 in follow-on release(s) so as to keep the changes brief enough for an effective code review.

## Dependencies

None

## Impact

None

This PR is part 1 for changes in #85.

### CODE REVIEW

**Comment by EdwardSafford-NOAA** [comment]
@CoryMartin-NOAA or @kevindougherty-noaa please assign reviewers to this PR.  Also a question -- it looks like the Application Test fails with the `testIodaObsSpaceAircraft.yaml` case.  Is that a known problem or did I just step in something?

**Comment by CoryMartin-NOAA** [comment]
@EdwardSafford-NOAA not sure why that is failing, looks like it is trying to download shapefiles... but odd it would start doing that now. I don't think it is something you did, we may need to investigate separately

**Comment by EdwardSafford-NOAA** [comment]
@CoryMartin-NOAA for what it's worth I was able to download that shape file (to my local machine), and it appears to be ok.  I wonder if the test's download just got interrupted.  I'd re-run the tests but I don't find a way to do that (I'm guessing that's another permission thing).

**Comment by CoryMartin-NOAA** [comment]
@EdwardSafford-NOAA good suggestion let me try that, yeah my guess is you have to be a 'maintainer' or whatever permission I have on this repo

**Comment by EdwardSafford-NOAA** [comment]
@CoryMartin-NOAA well, heck.  Seemed like a good idea, but it failed in the same place.

**Comment by CoryMartin-NOAA** [comment]
@EdwardSafford-NOAA I can confirm the same issue is happening re-running tests on develop... so this needs to be investigated. I'm helping organize a code sprint this week so not sure how much time I'll have to diagnose, will get back to you soon once I/others can find time

**Comment by EdwardSafford-NOAA** [comment]
FWIW I ran `testIodaObsSpaceAircraft.yaml` with both the develop and my PR branches.  Both run to completion without error.

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [inline src/eva/data/mon_data_space.py:300]: Can we at least combine these somehow where the if statement defines the empty size and dims that get reshaped? the for loops should (mostly, maybe need a dim=3 version) be able to be written only once

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [inline src/eva/data/mon_data_space.py]: Similar comment to above, I think these may be able to be consolidated a bit, but 4:15pm Friday brain doesn't have the exact solution

File: src/eva/utilities/utils.py
Comment (CoryMartin-NOAA) [inline src/eva/utilities/utils.py:341]: does this exist already in python somewhere maybe? or not?

Comment (EdwardSafford-NOAA) [reply src/eva/utilities/utils.py:341]: Seems like it should but I haven't found it so far.  This solution was a cut and paste from stackoverflow.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py:300]: Fair point.  I've made it a bit more concise.  The dim=3 case is still a special case, but maybe I can figure a way to make it less special.

File: src/eva/utilities/utils.py
Comment (CoryMartin-NOAA) [reply src/eva/utilities/utils.py:341]: works for me, just wanted to check

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [inline src/eva/data/mon_data_space.py:292]: What about printing a warning (or error) about missing filename using the logger?

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: I've got an idea here.  I think if I create a dictionary from the variable dimensions and coordinates I can then create the dataset via `xr.Dataset.from_dict(d)`.  That would put the if `num_dims` logic inside the for loop.  Giving it a shot now.

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [reply src/eva/data/mon_data_space.py:300]: Sounds like a plan, this is already more concise but if you can make it even better with your new plan, that would be great!

**Review by CoryMartin-NOAA — COMMENTED** [review]
One more comment, but after your next round I think this will be close to merging, thanks for the hard work so far!

File: src/eva/data/mon_data_space.py
Comment (kevindougherty-noaa) [inline src/eva/data/mon_data_space.py]: Logic is fine, a more concise way:
`x_range = channo if coords['xdef'] == 'Channel' else np.arange(1, dims['xdef']+1)`

File: src/eva/data/mon_data_space.py
Comment (kevindougherty-noaa) [inline src/eva/data/mon_data_space.py]: Same as above:
`y_range = channo if coords['ydef'] == 'Channel' else np.arange(1, dims['ydef']+1)`

File: src/eva/data/mon_data_space.py
Comment (kevindougherty-noaa) [inline src/eva/data/mon_data_space.py]: Could possibly use a dictionary like factory here:

```
d = {
1: np.reshape([[cycle_tm] * dims['xdef']], (dims['xdef'])),
2: np.reshape([[cycle_tm] * dims['xdef']] * dims['ydef'],
                      (dims['xdef'], dims['ydef'])),
3: np.reshape([[cycle_tm] * dims['xdef'] * dims['ydef'] * dims['zdef']],
                       (dims['xdef'], dims['ydef'], dims['zdef']))
}

try:
    cycle_arr = d[ndims_used]
except KeyError:
    self.logger.abort(f'ndims_used must be in range of 1-3, value is {ndims_used}')
```

I have not tested this, but could be a solution.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py:292]: I'll add a warning.  It's not uncommon for there to be missing files over a given time span.  Whatever scripting solution I/we use to create the yaml files should step over the gaps and give the yaml a clean list, but the bottom line is one missing file isn't a fatal condition.

**Review by kevindougherty-noaa — COMMENTED** [review]
Just a few suggestions I saw. Nothing that absolutely needs to be done, just could help add concise code.

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [reply src/eva/data/mon_data_space.py:292]: Ok that's fine but yes something like `logging.info(f"File {filename} is missing")` should be sufficient

Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py:292]: Done.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: See if you like this solution.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: I like it and have made it so.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: Done.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: It is a solution.  Cool!  It's about the same code length but I went this way so I'd remember the construct for future reference.

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [inline src/eva/data/mon_data_space.py:346]: this is nice!

File: src/eva/data/mon_data_space.py
Comment (CoryMartin-NOAA) [reply src/eva/data/mon_data_space.py]: looks good!

**Review by CoryMartin-NOAA — APPROVED** [review]
Thanks @EdwardSafford-NOAA ! I'll let @kevindougherty-noaa do another pass through but once he gives the :+1: I'll merge this in.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py:346]: Super!  Please note that this PR won't close #85.  It's step 1 of probably 2 PRs.  Thanks.

File: src/eva/data/mon_data_space.py
Comment (kevindougherty-noaa) [inline src/eva/data/mon_data_space.py]: Your logic is not wrong at all here, but I think I found a way to make this a bit more efficient. I used the factory-like method similar to what I suggested last review. It gets rid of all of the conditionals which are slow in Python. Please double check to make sure this works:

```
ndims_dict = {
    1: {
        'dims': (coords['xdef']),
        'cycle': Dataset({'cycle': ((coords['xdef']), cyc_darr),},
                         coords={coords['xdef']: np.arange(1, dims['xdef']+1)},)
    },
    2: {
        'dims': (coords['xdef'], coords['ydef']),
        'cycle': Dataset({'cycle': ((coords['xdef'], coords['ydef']), cyc_darr),},
                         coords={coords['xdef']: np.arange(1, dims['xdef']+1),
                         coords['ydef']: np.arange(1, dims['ydef']+1)},)
    },
    3: {
        'dims': (coords['xdef'], coords['ydef'], coords['zdef']),
        'cycle': Dataset({'cycle': ((coords['xdef'], coords['ydef'], coords['zdef']), cyc_darr),},
                         coords={coords['xdef']: np.arange(1, dims['xdef']+1),
                         coords['ydef']: np.arange(1, dims['ydef']+1),
                         coords['zdef']: np.arange(1, dims['zdef']+1)},)
    }
}

# Create empty dict
d = {}

# Loop through vars and fill empty dict
for x in range(0, nvars):
    d[vars[x]] = {
        'dims': ndims_dict[ndims_used]['dims'],
        'data': darr[x, :, :]
    }

# Create Dataset from dictionary
rtn_ds = Dataset.from_dict(d)

# Merge cycle Dataset based on ndims_used
rtn_ds = rtn_ds.merge(ndims_dict[ndims_used]['cycle'])
```

**Review by kevindougherty-noaa — COMMENTED** [review]
I added a suggestion that may save some lines of code and helps get rid of multiple conditionals.

File: src/eva/data/mon_data_space.py
Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: Cool.  I had tried doing that here before but, looking at your code, I see where I might have gone wrong in that attempt.  I'll give it another go.

Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: @kevindougherty-noaa I'm getting this error with the 'cycle' definition in the factory-like construct:

`ValueError: Could not convert tuple of form (dims, data[, attrs, 
encoding]): ('Channel', array([[[datetime.datetime(2023, 4, 9, 0, 0), .... to Variable`

Any thoughts on how to correct that?

Comment (kevindougherty-noaa) [reply src/eva/data/mon_data_space.py]: @EdwardSafford-NOAA I am not too sure what you mean. Is it in issue in the dictionary or when you try and merge?

Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: It's in the dictionary.  I've tried paring down/simplifying the dictionary structure.  The 'dims' definition is fine.  The 'cycle' is the source of the problem.

Comment (kevindougherty-noaa) [reply src/eva/data/mon_data_space.py]: Maybe `Dataset()` can't work in the dictionary. I unfortunately don't have a solution for the error you are getting.

Comment (EdwardSafford-NOAA) [reply src/eva/data/mon_data_space.py]: I worked my way around the `Dataset `issue.  This isn't quite as concise as what you'd hoped for but all the the `if/elif/else `on `num_dims `is now eliminated.  Please see what you think.

**Review by kevindougherty-noaa — APPROVED** [review]
Looks great. I appreciate your patience and willingness to try the suggested changes!

**Comment by EdwardSafford-NOAA** [comment]
Thanks very much for your suggestions.  It's always great to learn new things!

**Pull request merged**
